jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -0,0 +1,570 @@
+import java.lang.invoke.MethodHandles;
+import static java.lang.invoke.MethodHandles.*;
+import java.lang.invoke.MethodHandles.Lookup;
+import java.lang.invoke.MethodType;
+import static java.lang.invoke.MethodType.methodType;
+import java.lang.invoke.WrongMethodTypeException;
+
+import jdk.incubator.foreign.FunctionDescriptor;
+import jdk.incubator.foreign.ValueLayout;
+import jdk.incubator.foreign.MemoryLayout;
+import jdk.incubator.foreign.Addressable;
+import jdk.incubator.foreign.MemoryAddress;
+import jdk.incubator.foreign.MemorySegment;
+import jdk.incubator.foreign.LibraryLookup;
+import static jdk.incubator.foreign.LibraryLookup.Symbol;
+import jdk.incubator.foreign.CLinker.TypeKind;
+import static jdk.incubator.foreign.CLinker.TypeKind.*;
+
+/**
+ * The counterpart in OpenJDK is replaced with this class that wrap up a method handle
+ * enabling the native code to the ffi_call via the libffi interface at runtime.
+ */
+public class ProgrammableInvoker {
+	
+	private final MethodType funcMethodType;
+	private final FunctionDescriptor funcDescriptor;
+	private final Addressable functionAddr;
+	private long cifNativeThunkAddr;
+	private long argTypesAddr;
+	private List<MemoryLayout> argLayouts;
+	private MemoryLayout[] argLayoutArray;
+	private MemoryLayout realReturnLayout;
+	
+	final static Lookup lookup = MethodHandles.lookup();
+	
+	/* The prep_cif and the corresponding argument layouts are cached & shared in multiple downcalls/threads */
+	private static final HashMap<Integer, Long> cachedCifNativeThunkAddr = new HashMap<Integer, Long>();
+	private static final HashMap<List<MemoryLayout>, Long> cachedArgLayouts = new HashMap<List<MemoryLayout>, Long>();
+
+	/* Argument filters that convert the primitive types or MemoryAddress to long */
+	private static MethodHandle booleanToLongArgFilter = null;
+	private static MethodHandle charToLongArgFilter = null;
+	private static MethodHandle byteToLongArgFilter = null;
+	private static MethodHandle shortToLongArgFilter = null;
+	private static MethodHandle intToLongArgFilter = null;
+	private static MethodHandle floatToLongArgFilter = null;
+	private static MethodHandle doubleToLongArgFilter = null;
+	private static MethodHandle memAddrToLongArgFilter = null;
+	
+	/* Return value filters that convert the Long object to the primitive types or MemoryAddress */
+	private static MethodHandle longObjToVoidRetFilter = null;
+	private static MethodHandle longObjToBooleanRetFilter = null;
+	private static MethodHandle longObjToCharRetFilter = null;
+	private static MethodHandle longObjToByteRetFilter = null;
+	private static MethodHandle longObjToShortRetFilter = null;
+	private static MethodHandle longObjToIntRetFilter = null;
+	private static MethodHandle longObjToLongRetFilter = null;
+	private static MethodHandle longObjToFloatRetFilter = null;
+	private static MethodHandle longObjToDoubleRetFilter = null;
+	private static MethodHandle longObjToMemAddrRetFilter = null;
```

Comment:
Initializing all these fields is at best redundant. Can they be made `final` (they all appear to be initialized in `<clinit>`).

Reply:
Fixed as suggested above.